torchmdnet/models/tensornet.py

```diff
@@ -0,0 +1,397 @@
+import torch
+import numpy as np
+from typing import Optional, Tuple
+from torch import Tensor, nn
+from torch_scatter import scatter
+from torch_geometric.nn import MessagePassing
+from torchmdnet.models.utils import (
+    CosineCutoff,
+    Distance,
+    rbf_class_mapping,
+    act_class_mapping,
+)
+
+# Creates a skew-symmetric tensor from a vector
+def vector_to_skewtensor(vector):
+    tensor = torch.cross(*torch.broadcast_tensors(vector[...,None], torch.eye(3,3, device=vector.device)[None,None]))
+    return tensor.squeeze(0)
+
+# Creates a symmetric traceless tensor from the outer product of a vector with itself
+def vector_to_symtensor(vector):
+    tensor = torch.matmul(vector.unsqueeze(-1), vector.unsqueeze(-2))
+    I = (tensor.diagonal(offset=0, dim1=-1, dim2=-2)).mean(-1)[...,None,None] * torch.eye(3,3, device=tensor.device)
+    S = 0.5 * (tensor + tensor.transpose(-2,-1)) - I
+    return S
+
+# Full tensor decomposition into irreducible components
+def decompose_tensor(tensor):
+    I = (tensor.diagonal(offset=0, dim1=-1, dim2=-2)).mean(-1)[...,None,None] * torch.eye(3,3, device=tensor.device)
+    A = 0.5 * (tensor - tensor.transpose(-2,-1))
+    S = 0.5 * (tensor + tensor.transpose(-2,-1)) - I
+    return I, A, S
+
+# Modifies tensor by multiplying invariant features to irreducible components
+def new_radial_tensor(I, A, S, f_I, f_A, f_S):
+    I = (f_I)[...,None,None] * I
+    A = (f_A)[...,None,None] * A
+    S = (f_S)[...,None,None] * S
+    return I, A, S
+
+# Computes Frobenius norm
+def tensor_norm(tensor):
+    return (tensor**2).sum((-2,-1))
+
+ 
+class TensorNet(nn.Module):
+    def __init__(
+        self,
+        hidden_channels=128,
+        num_layers=2,
+        num_rbf=32,
+        activation="silu",
+        rbf_type = 'expnorm',
+        cutoff_lower=0,
+        cutoff_upper=4.5,
+        max_num_neighbors=64,
+        return_vecs=True,
+        loop=True,
+        trainable_rbf=False,
+        max_z = 128,
+        equivariance_invariance_group = "O(3)",
+    ):
+        super(TensorNet, self).__init__()
+        
+        assert rbf_type in rbf_class_mapping, (
```

Comment:
Lets add some documentation similar to ET:
https://github.com/torchmd/torchmd-net/blob/237b4fe1bf0ce52721b42fe51f35efca018753f9/torchmdnet/models/torchmd_et.py#L16-L53
In this case a reference to the paper will also be cool.